protocol/x/clob/keeper/deleveraging.go

```diff
@@ -499,23 +529,50 @@ func (k Keeper) ProcessDeleveraging(
 		),
 	)
 
-	// Send on-chain update for the deleveraging. The events are stored in a TransientStore which should be rolled-back
-	// if the branched state is discarded, so batching is not necessary.
-	k.GetIndexerEventManager().AddTxnEvent(
-		ctx,
-		indexerevents.SubtypeDeleveraging,
-		indexerevents.DeleveragingEventVersion,
-		indexer_manager.GetBytes(
-			indexerevents.NewDeleveragingEvent(
-				liquidatedSubaccountId,
-				offsettingSubaccountId,
-				perpetualId,
-				satypes.BaseQuantums(new(big.Int).Abs(deltaBaseQuantums).Uint64()),
-				satypes.BaseQuantums(deltaQuoteQuantums.Uint64()),
-				deltaBaseQuantums.Sign() > 0,
-			),
-		),
-	)
+	return nil
+}
+
+// DeleverageSubaccountsInFinalSettlementMarkets uses the subaccountOpenPositionInfo returned from the
+// liquidations daemon to deleverage subaccounts with open positions in final settlement markets. Note
+// this function will deleverage both negative TNC and non-negative TNC subaccounts. The deleveraging code
+// uses the bankruptcy price for the former and the oracle price for the latter.
+func (k Keeper) DeleverageSubaccountsInFinalSettlementMarkets(
+	ctx sdk.Context,
+	subaccountOpenPositionInfo map[uint32]map[bool]map[satypes.SubaccountId]struct{},
+	numDeleveragingAttempts int,
+) error {
+	// Gather perpetualIds for perpetuals whose clob pairs are in final settlement.
+	finalSettlementPerpetualIds := make(map[uint32]struct{})
+	for _, clobPair := range k.GetAllClobPairs(ctx) {
+		if clobPair.Status == types.ClobPair_STATUS_FINAL_SETTLEMENT {
+			perpetualId := clobPair.MustGetPerpetualId()
+			finalSettlementPerpetualIds[perpetualId] = struct{}{}
+		}
+	}
+
+	// Deleverage subaccounts with open positions in final settlement markets.
+	for perpetualId := range finalSettlementPerpetualIds {
+		for isBuy := range subaccountOpenPositionInfo[perpetualId] {
+			for subaccountId := range subaccountOpenPositionInfo[perpetualId][isBuy] {
+				if numDeleveragingAttempts >= int(k.Flags.MaxDeleveragingAttemptsPerBlock) {
```

Comment:
Should we use a separate flag to limit the number of final settlement deleveraging ops? This looks through all subaccounts with open positions in the market so it's bounded right?

Reply:
I thought about this, but opted for re-using the flag in the interest of respecting the current performance implications of proposing deleveraging operations.

When you say this operation is bounded, what do you mean by that? In theory there is no limit to the number of subaccounts that can havae open positions in a market so it isn't bounded